internal/oci/runtime_oci.go

```diff
@@ -1220,3 +1231,248 @@ func (r *Runtime) SpoofOOM(c *Container) {
+	// wants to connect to the main CRIU process outside of the container.
+	// After changing the default label of new sockets during checkpointing
+	// The original value needs to be restored.
+	socketLabel, err := selinux.SocketLabel()
+	// The default socket label returns EOF when reading it.
+	// This could be seen as a bug in go-selinux.
+	// When we later use socketLabel to reset the label to default
+	// we could just use an empty string, but first reading it is probably
+	// more correct (and maybe useless if CRI-O never uses socket
+	// specific SELinux labels).
+	if err != nil && err != io.EOF && selinux.GetEnabled() {
+		return errors.Wrapf(err, "Reading default socket label failed")
+	}
+
+	// workPath will be used to store dump.log and stats-dump
+	workPath := c.Dir()
+	// imagePath is used by CRIU to store the actual checkpoint files
+	imagePath := c.CheckpointPath()
+
+	logrus.Debugf("Writing checkpoint to %s", imagePath)
+	logrus.Debugf("Writing checkpoint logs to %s", workPath)
+	args := []string{}
+	args = append(
+		args,
+		"--criu",
+		r.config.CriuPath,
+		rootFlag,
+		r.root,
+		"checkpoint",
+		"--image-path",
+		imagePath,
+		"--work-path",
+		workPath,
+	)
+	if leaveRunning {
+		args = append(args, "--leave-running")
+	}
+
+	args = append(args, c.id)
+
+	if selinux.GetEnabled() {
+		// Change our own socket label to match the own from the container
+		// process so that the parasite can connect from within the container
+		// to the CRIU process outside of the container.
+		if err := selinux.SetSocketLabel(specgen.Process.SelinuxLabel); err != nil {
+			return errors.Wrapf(err, "Cannot set socket label for container %q to %q", c.ID(), specgen.Process.SelinuxLabel)
+		}
+		// Whatever happens now, it is important to not exit this function
+		// without resetting the socket label to default.
+
+		// Also, if the Go runtime creates a thread between setting and resetting
+		// the socket label it might run with the wrong SELinux label.
+		// It would probably be good to block the creation of new threads here.
+	}
+
+	_, err = utils.ExecCmd(r.path, args...)
+
+	// Reset socket label to default.
+	// This is not happening in Podman as Podman exits after checkpointing.
+	// Not doing as a defer function as this needs to run as soon as possible after running runc.
+	if labelErr := selinux.SetSocketLabel(socketLabel); labelErr != nil {
+		if selinux.GetEnabled() {
+			return errors.Wrapf(labelErr, "Cannot reset socket label to original value (%s)", socketLabel)
+		}
+	}
+
+	if err != nil {
+		return errors.Wrapf(err, "Running %q %q failed", r.path, args)
+	}
+
+	if !leaveRunning {
+		c.state.Status = ContainerStateStopped
+		c.state.ExitCode = utils.Int32Ptr(0)
+		c.state.Finished = time.Now()
+	}
+
+	return nil
+}
+
+// RestoreContainer restores a container.
+func (r *runtimeOCI) RestoreContainer(c *Container, sbSpec *rspec.Spec, infraPid int, cgroupParent string) error {
+	c.opLock.Lock()
```

Comment:
Here is a potential deadlock. In function `RestoreContainer()`, it will do `c.opLock.Lock()` then call `CreateContainer()` at line 1449 to restore container. If restore failed(like criu returns errors), it will call `DeleteContainer()` at line 200 to delete the container and it will try to do `c.opLock.Lock()` again , hence causes a deadlock.